libqtile/widget/bluetooth.py

```diff
@@ -38,32 +40,105 @@ class Bluetooth(base.InLoopPollText):
     ]
 
     def __init__(self, **config):
-        base.InLoopPollText.__init__(self, **config)
+        base._TextBox.__init__(self, '', **config)
         self.add_defaults(Bluetooth.defaults)
 
-        bus = dbus.SystemBus()
-        # set up interface into adapter properties
-        adapter = bus.get_object('org.bluez', '/org/bluez/hci0')
-        adapter_interface = dbus.Interface(adapter, 'org.bluez.Adapter1')
-        self._adapter = dbus.Interface(adapter_interface, 'org.freedesktop.DBus.Properties')
-        # set up interface into device properties
-        device = bus.get_object('org.bluez', '/org/bluez/hci0/' + config['hci'])
-        device_interface = dbus.Interface(device, 'org.bluez.Device1')
-        self._device = dbus.Interface(device_interface, 'org.freedesktop.DBus.Properties')
-
-    def poll(self):
-        try:
-            powered = self._adapter.Get('org.bluez.Adapter1', 'Powered')
-            if powered == 0:
-                status = 'off'
-            else:
-                connected = self._device.Get('org.bluez.Device1', 'Connected')
-                if connected == 0:
-                    status = 'on'
-                else:
-                    status = self._device.Get('org.bluez.Device1', 'Name')
+        self.hci = config['hci']
+
+    async def _config_async(self):
+        # get initial device name
+        _, device_msg = await _send_dbus_message(session_bus=False,
```

Comment:
This isn't really the correct way to retrieve properties. The `_send_dbus_message` method is a low-level method used by a couple of helper functions.

Have a look at https://python-dbus-next.readthedocs.io/en/latest/high-level-client/index.html for an example of how the client connection should work.

Reply:
Have a look at this widget that I've been working on (https://github.com/elParaguayo/qtile/blob/widget-laptopbattery/libqtile/widget/laptopbattery.py#L179). It gives an example of how to get dbus object properties.

Reply:
Have a look at this widget I've been working on https://github.com/elParaguayo/qtile/blob/widget-laptopbattery/libqtile/widget/laptopbattery.py#L179. It gives an example of how to get dbus object properties.

NB this was made before several changes to my dbus-next pull request. `_config_async` is now a coroutine so the contents of my widget's `_setup_dbus` will need to be moved into that method. I just haven't got round to doing it yet.

Reply:
No worries, wanted to make sure people had access to this by the next release.

Reply:
No problem. I imagine tyere'll be lots of time.

Reply:
Take a look at my latest commit and let me know if that's more what you were thinking.